tests/framework/framework.go

```diff
@@ -0,0 +1,242 @@
+})
+fmt.Printf("\n****** lookup: kubectl-path=%q\n", flag.Lookup("kubectl-path").Value)
+		if *kubectlPath == "" {
+			Fail("flag `kubectl-path` was not specified")
+		}
+		f.KubectlPath = *kubectlPath
+		if *cdiInstallNs == "" {
+			Fail("flag `installed-namespace` was not specified")
+		}
+		f.CdiInstallNs = *cdiInstallNs
+		if *kubeConfig == "" {
+			Fail("flag `kubeconfig` was not specified")
+		}
+		f.KubeConfig = *kubeConfig
+		if *master == "" {
+			Fail("flag `master` was not specified")
+		}
+		f.Master = *master
+fmt.Printf("\n******framework:\nf.KubectlPath=%q, f.CdiInstallNs=%q, f.KubeConfig=%q, f.Master=%q",f.KubectlPath, f.CdiInstallNs, f.KubeConfig, f.Master)
+	}
+
+	BeforeEach(f.BeforeEach)
+	AfterEach(f.AfterEach)
+
+	return f
+}
+
+func (f *Framework) BeforeEach() {
+	// generate unique primary ns (ns2 not created here)
+	By(fmt.Sprintf("Building a %q namespace api object", f.NsPrefix))
+	ns, err := f.CreateNamespace(f.NsPrefix, map[string]string{
+		"cdi-e2e": f.NsPrefix,
+	})
+	Expect(err).NotTo(HaveOccurred())
+	f.Namespace = ns
+	f.AddNamespaceToDelete(ns)
+
+	// clients
+	if f.K8sClient == nil {
+		By("Creating a kubernetes client")
+		kcs, err := f.GetKubeClient()
+		Expect(err).NotTo(HaveOccurred())
+		f.K8sClient = kcs
+	}
+	if f.CdiClient == nil {
+		By("Creating a CDI client")
+		cs, err := f.GetCdiClient()
+		Expect(err).NotTo(HaveOccurred())
+		f.CdiClient = cs
+	}
+}
+
+func (f *Framework) AfterEach() {
+	// delete the namespace(s) in a defer in case future code added here could generate
+	// an exception. For now there is only a defer.
+	defer func() {
+		for _, ns := range f.namespacesToDelete {
+			if len(ns.Name) == 0 {
+				continue
+			}
+			By(fmt.Sprintf("Destroying namespace %q for this suite.", ns.Name))
+			err := DeleteNS(f.K8sClient, ns.Name)
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}()
+	return
+}
+
+// Instantiate a new namespace object with a unique name and the passed-in label(s).
+func (f *Framework) CreateNamespace(prefix string, labels map[string]string) (*v1.Namespace, error) {
+	ns := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("cdi-e2e-tests-%s-", prefix),
+			Namespace:    "",
+			Labels:       labels,
+		},
+		Status: v1.NamespaceStatus{},
+	}
+
+fmt.Printf("\n******* ns spec=%+v\n", ns)
+	var nsObj *v1.Namespace
+	c := f.K8sClient
+	err := wait.PollImmediate(2*time.Second, NsCreateTime, func() (bool, error) {
+		var err error
+		nsObj, err = c.CoreV1().Namespaces().Create(ns)
+		if err != nil {
+			glog.Errorf("Unexpected error while creating %q namespace: %v", ns.GenerateName, err)
+			return false, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+fmt.Printf("\n******* nsObj=%+v\n", nsObj)
+	return nsObj, nil
+}
+
+func (f *Framework) AddNamespaceToDelete(ns *v1.Namespace) {
+	f.namespacesToDelete = append(f.namespacesToDelete, ns)
+}
+
+func DeleteNS(c *kubernetes.Clientset, ns string) error {
```

Comment:
Any reason this isn't part of the Framework, since we are automatically cleaning up namespaces in the AfterEach, I wouldn't except people to manually call DeleteNS, I might even not export it.

Reply:
Actually found my answer, DestroyAllTestNamespaces uses it, so you can ignore my previous comment

Reply:
We should converge on a strategy for deleting test namespaces. I like your label filter idea (but it can be implemented more efficiently, I think). But I'd like the clean up tied to the framework too.

Reply:
I think we should do both, which means that most of the time the After/BeforeSuite does nothing. But if the framework fails for whatever reason, we have the label as a backup.

Reply:
per discussion with @awels we will not define an `AfterSuite` to delete all cdi labeled namespaces. The reason is that it could be possible for multiple cdi tests _suites_ to be running in parallel. If we delete all namespaces with the cdi label we could be deleting active namespaces. Better to safe.